Add support for Dwarf 5 in MCCAS

rastogishubham opened the pull request:
MCCAS doesn't support DWARF 5 currently. Since we are switching to DWARF 5 on all platforms relatively soon, we need to have MCCAS support it as well.

rastogishubham committed:
Add support for Dwarf 5 Line table header parsing in MCCAS

rastogishubham committed:
Add RawSValue() to DWARFFormValue and writeSLEB128 to DataWriter

To support reading and writing DW_FORM_implicit_const in MCCAS, there is
a need to be able to read a signed value of a DWARFFormValue and be able
to write a signed LEB128 value to a CASObject. This change adds those
two functions to their respective classes.

rastogishubham committed:
Add PaddingRef after debug_str section

In DWARF 5, we can have padding after the debug_str section, so a
PaddingRef must be added after the CAS blocks for individual strings.

rastogishubham commented:
@swift-ci please test llvm

adrian-prantl commented on llvm/include/llvm/MCCAS/MCCASDebugV1.h:
```diff
@@ -72,6 +72,9 @@ struct DataWriter {
   /// Write ULEB128(V) to the data stream.
   void writeULEB128(uint64_t V) { encodeULEB128(V, DataStream); }
 
+  /// Write ULEB128(V) to the data stream.
```
SLEB

adrian-prantl commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -743,9 +743,16 @@ getLineTableLengthInfoAndVersion(DWARFDataExtractor &LineTableDataReader,
-    return createStringError(inconvertibleErrorCode(),
-                             "DWARF 5 and above is not currently supported");
+  if (Version >= 5) {
+    // DWARF v5 6.2.4 read address_size and segment_selector_size
```
I'm having trouble parsing this sentence...

adrian-prantl commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -743,9 +743,16 @@ getLineTableLengthInfoAndVersion(DWARFDataExtractor &LineTableDataReader,
-                             "DWARF 5 and above is not currently supported");
+  if (Version >= 5) {
+    // DWARF v5 6.2.4 read address_size and segment_selector_size
+    LineTableDataReader.getU8(OffsetPtr, &Err);
```
Since you are discarding the Address size, shouldn't you at least error out if it isn't 64 bit, which I'm assuming you are implicitly assuming?

adrian-prantl commented on llvm/include/llvm/MCCAS/MCCASObjectV1.h:
```diff
@@ -580,6 +581,7 @@ class MCCASBuilder {
   struct CUSplit {
     SmallVector<MutableArrayRef<char>> SplitCUData;
     SmallVector<size_t> AbbrevOffsets;
+    SmallVector<uint16_t> DwarfVersions;
```
What's the contents of this vector?

adrian-prantl commented on llvm/lib/MCCAS/MCCASDebugV1.cpp:
```diff
@@ -255,6 +255,14 @@ void AbbrevEntryWriter::writeAbbrevEntry(DWARFDie DIE) {
+    // attributes with this form, the attribute specification contains a third
+    // part, which is a signed LEB128 number. The value of this number is used
+    // as the value of the attribute, and no value is stored in the .debug_info
+    // section.
```
I think you may need to paraphrase the text here since the text you're quoting is under a license that may not be compatible with the LLVM one.

rastogishubham commented on llvm/test/DebugInfo/CAS/AArch64/debug_unopt.ll:
```diff
@@ -2,6 +2,7 @@
 ; CHECK: mc:debug_string_section llvmcas://
 ; CHECK-NEXT: mc:debug_string llvmcas://
+; CHECK-NEXT: mc:padding      llvmcas://
 ; CHECK-NEXT: mc:section      llvmcas://
```
In my commit refactoring, I didn't realize this change should be part of the previous commit

rastogishubham commented on llvm/test/tools/llvm-cas-dump/basic_debug_test.ll:
```diff
@@ -43,6 +43,7 @@
 ; CHECK-NEXT:       mc:debug_string  llvmcas://{{.*}}
 ; CHECK-NEXT:       mc:debug_string  llvmcas://{{.*}}
 ; CHECK-NEXT:       mc:debug_string  llvmcas://{{.*}}
+; CHECK-NEXT:       mc:padding  llvmcas://{{.*}}
```
This one too

rastogishubham commented on llvm/test/DebugInfo/CAS/AArch64/debug_unopt.ll:
```diff
@@ -2,6 +2,7 @@
 ; RUN: llvm-cas-dump --cas=%t/cas --casid-file %t/debug_unopt.id | FileCheck %s
 ; CHECK: mc:debug_string_section llvmcas://
 ; CHECK-NEXT: mc:debug_string llvmcas://
+; CHECK-NEXT: mc:padding      llvmcas://
```
In my commit refactoring, I didn't realize this change should be part of the previous commit

rastogishubham committed:
Add support for reading and writing DW_FORM_implicit_const in MCCAS

In DWARF 5, there is a new form, DW_FORM_implicit_const, which encodes
the value of the attribute in the debug_abbrev section, as a signed
LEB128 number. This change adds support for being able to parse that
and write it into MCCAS, as well as read it from a CASObject.

rastogishubham commented:
@swift-ci please test llvm

rastogishubham commented:
@swift-ci please test llvm

rastogishubham committed:
Pass debug_str_offsets section to InMemoryCASDWARFObject class

In Dwarf 5, for proper UnitDIE extraction, the debug_str_offsets section
needs to be passed to the DWARFObject. This patch adds support for that.

rastogishubham committed:
Add support for Dwarf 5 debug_info section in MCCAS

This includes, parsing the Dwarf 5 Compile Unit header correctly
and correctly creating and materializing from the CAS with the larger
DWARFv5 header.

rastogishubham commented:
@swift-ci please test

adrian-prantl commented on llvm/include/llvm/MCCAS/MCCASObjectV1.h:
```diff
@@ -580,6 +581,9 @@ class MCCASBuilder {
     SmallVector<MutableArrayRef<char>> SplitCUData;
     SmallVector<size_t> AbbrevOffsets;
+    // A list of Dwarf versions of the compile units, to help determine what the
+    // Compile Unit header format and size is like when creating CASObjects.
```
///

felipepiovezan commented on llvm/lib/MCCAS/MCCASDebugV1.cpp:
```diff
@@ -330,6 +351,18 @@ mccasformats::v1::reconstructAbbrevSection(raw_ostream &OS,
+      if (Form == dwarf::Form::DW_FORM_implicit_const) {
+        int64_t ImplicitVal;
+        if (auto Err = Reader.readSLEB128(ImplicitVal))
+          handleAllErrors(std::move(Err));
```
the duplication of this block with the one above is a bit concerning; is there a way to factor this into a helper function?

felipepiovezan reviewed (approved):
LGTM but please see the one comment I left!

rastogishubham commented:
@swift-ci please test llvm
